Firestore/core/src/api/load_bundle_task.h

```diff
@@ -0,0 +1,197 @@
+         lhs.bytes_loaded() == rhs.bytes_loaded() &&
+         lhs.documents_loaded() == rhs.documents_loaded() &&
+         lhs.total_bytes() == rhs.total_bytes() &&
+         lhs.total_documents() == rhs.total_documents();
+}
+
+inline bool operator!=(const LoadBundleTaskProgress lhs,
+                       const LoadBundleTaskProgress& rhs) {
+  return !(lhs == rhs);
+}
+
+/** A handle used to lookup and remove observer from the task. */
+using LoadBundleHandle = std::string;
+
+/** Observer type that is called by the task when there is an update. */
+using ProgressObserver = std::function<void(LoadBundleTaskProgress)>;
+
+/** Holds the `LoadBundleHandle` to `ProgressObserver` mapping. */
+using HandleObservers =
+    std::vector<std::pair<LoadBundleHandle, ProgressObserver>>;
+
+/**
+ * Represents the task of loading a Firestore bundle. It provides progress of
+ * bundle loading, as well as task completion and error events.
+ */
+class LoadBundleTask {
+ public:
+  explicit LoadBundleTask(std::shared_ptr<util::Executor> user_executor)
+      : user_executor_(std::move(user_executor)) {
+  }
+
+  /**
+   * Instructs the task to notify the specified observer when there is a
+   * progress update with the given `LoadBundleTaskState`.
+   *
+   * @return A handle that can be used to remove the callback from this task.
+   */
+  LoadBundleHandle ObserveState(LoadBundleTaskState state,
+                                ProgressObserver callback);
+
+  /**
+   * Removes the observer associated with the given handle, does nothing if the
+   * callback cannot be found.
+   */
+  void RemoveObserver(LoadBundleHandle);
+
+  /**
+   * Removes all observers associated with the given `LoadBundleTaskState`.
+   */
+  void RemoveObservers(LoadBundleTaskState state);
+
+  /** Removes all observers. */
+  void RemoveAllObservers();
+
+  /**
+   * Notifies observers with a success progress. Both `Success` and `InProgress`
+   * observers will get notified.
+   */
+  void SetSuccess(LoadBundleTaskProgress success_progress);
+
+  /**
+   * Notifies observers with a error progress, by changing the last progress
+   * this instance has been with an `Error` state.
+   *
+   * Both `Error` and `InProgress` observers will get notified.
+   */
+  void SetError();
+
+  /** Notifies observers with a `InProgress` progress. */
+  void UpdateProgress(LoadBundleTaskProgress progress);
+
+ private:
+  /** Gets all observers associated with the given state. */
+  HandleObservers& GetObservers(LoadBundleTaskState state);
+
+  /** Notifies all given observers. */
+  void ExecuteCallbacks(const HandleObservers& observers);
+
+  /** The executor to run all observers when notified. */
+  std::shared_ptr<util::Executor> user_executor_;
+
+  /** Guard to all internal state mutation. */
+  mutable std::mutex mutex_;
+
+  /** An array holds mapping from `LoadBundleTaskState` values to observers. */
+  std::array<HandleObservers, 3> observers_by_states_;
```

Comment:
But that would make places handing the enum awkward. Would defining a const value for count help?

Reply:
> But that would make places handing the enum awkward.

Can you give an example?

> Would defining a const value for count help?

I think it would be better (though unlike an enum member, it could theoretically get out of sync).

Reply:
No longer relevant.